Add Sobol' sampler from SciPy

tupui opened the pull request:
Closes #486, closes #450 and should close #478 as if the method does work better (which I am not convinced yet from my tests) should go in SciPy directly.

Following discussions in #486, this adds Sobol' sampler as a new sampler and deprecate `salib.sample.saltelli`.

Note that SciPy also has LHS (and Halton that can be used otherwise).

This needs SciPy to 1.7 and doing so I updated NumPy and Python minimal versions following NEP29.

This is a draft as I need some guidance on how to proceed next. E.g. I removed the skipping as we can scramble and should have better properties doing so.

cc @ConnectedSystems

tupui commented:
I tried to update the ubuntu version in the CI but it would need more work. So probably better not to touch this here.

ConnectedSystems commented:
Thanks @tupui 

I will take a closer look when I can (may take a week or two) but at first glance this looks like a very clean re-implementation.

Re skipping vs scrambling, I would prefer to support skipping anyway, as I think the user should be able to decide what to apply. 

One thing I would like to see is expanded docstrings with more complete notes to cover the differences (I can do this if you like).

tupui commented:
Great thanks @ConnectedSystems. It took me quite some time to do this as well 😉

> I will take a closer look when I can (may take a week or two) but at first glance this looks like a very clean re-implementation.

It's highly based on the current Saltelli sampler. I just removed the skipping part.

> Re skipping vs scrambling, I would prefer to support skipping anyway, as I think the user should be able to decide what to apply.

But what is the use case? As seen in #518. Skipping bring difficulties if one want's to resample. And it's also a complex logic to keep the properties.

Instead, if the use case is to have different samples, I would suggest adding a `seed` or `random_state` parameter. This way users can increment this instead and the scrambler does the correct thing behind the scene.

This brings up the `scramble` parameter. I added this almost just as a provision in case you want to set a different strategy, but as it is, it should always be True. There is no operational reasons to not scramble the sequence.

> One thing I would like to see is expanded docstrings with more complete notes to cover the differences (I can do this if you like).

Sure, I can do this once we settle on the API.

ConnectedSystems commented:
> But what is the use case? As seen in https://github.com/SALib/SALib/issues/518. Skipping bring difficulties if one want's to resample. And it's also a complex logic to keep the properties.

I understand the reservation - my view is that SALib is useful as a teaching tool as well as a research tool. Someone out there may be relying on it for some purpose (e.g., to show case different strategies). The code itself is useful if one were interested in seeing how skipping might be implemented.

> Instead, if the use case is to have different samples, I would suggest adding a `seed` or `random_state` parameter

I think `seed` is best as it matches implementation of other methods.

tupui commented:
Ok I can add back the skip parameter following what's in the current Saltelli method. And I will add seed.

ConnectedSystems commented on src/SALib/sample/sobol.py:
```diff
@@ -0,0 +1,234 @@
+    qrng = qmc.Sobol(d=2*D, scramble=scramble, seed=seed)
+
+    # fast-forward logic
+    if skip_values > 0:
```
Is it acceptable if `skip_values > 0` and `scramble == True`?

ConnectedSystems replied:
Whatever the answer, it would be worthwhile to add it to the docstring I think

tupui replied:
Yes we can combine both. I added a note in the docstrings as suggested.

ConnectedSystems commented on src/SALib/sample/sobol.py:
```diff
@@ -0,0 +1,234 @@
+        saltelli_sequence = np.zeros([(2 * Dg + 2) * N, D])
+    else:
+        saltelli_sequence = np.zeros([(Dg + 2) * N, D])
+    index = 0
```
Please add line between `if` block and `index = 0` for clarity

tupui replied:
Done

ConnectedSystems commented on src/SALib/sample/sobol.py:
```diff
@@ -0,0 +1,234 @@
+        saltelli_sequence = np.zeros([(Dg + 2) * N, D])
+    index = 0
+
+    for i in range(N):
```
Not completely necessary but I suggest we replace these nested loops with `itertools.product()`

tupui replied:
I am not sure how to go about this. The loops need to update `index` in the first level. Maybe this can be kept for latter. And actually there might be another way to write these with NumPy operations.

ConnectedSystems reviewed (commented):
Thanks, just some minor changes and a question for you.

tupui reviewed (commented):
Thanks for the review @ConnectedSystems. I updated the PR

tupui commented:
@ConnectedSystems the CI failure is also present on `main`. I propose to fix this in another PR as I update the CI to use GitHub Actions.

tupui commented:
I updated the PR with the recent changes in `main`. Dependencies are already up to date now.
